File: course_4/module_4/programming_assignment_4/solution.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

// Graph represents a directed graph.
// It contains adjacency lists for both the original and reversed graphs.
type Graph struct {
	adj    map[int][]int // Adjacency list for the original graph.
	adjInv map[int][]int // Adjacency list for the reversed graph.
	nodes  int           // Total number of nodes in the graph.
}

// NewGraph initializes a new Graph object with the given number of nodes.
func NewGraph(nodes int) *Graph {
	return &Graph{
		adj:    make(map[int][]int),
		adjInv: make(map[int][]int),
		nodes:  nodes,
	}
}

// AddEdge adds a directed edge to the original graph and its reverse to the reversed graph.
func (g *Graph) AddEdge(u, v int) {
	g.adj[u] = append(g.adj[u], v)       // Add edge to the original graph.
	g.adjInv[v] = append(g.adjInv[v], u) // Add reversed edge to the reversed graph.
}

// SATSolver encapsulates the logic for solving the 2-SAT problem.
// It uses Kosaraju's algorithm to find strongly connected components (SCCs).
type SATSolver struct {
	graph        *Graph       // The graph representing the problem.
	visited      map[int]bool // Visited array for the original graph.
	visitedInv   map[int]bool // Visited array for the reversed graph.
	scc          map[int]int  // Maps each node to its SCC ID.
	stack        []int        // Stack used for Kosaraju's algorithm.
	counter      int          // Counter for assigning SCC IDs.
	numVariables int          // Number of variables in the 2-SAT problem.
}

// NewSATSolver initializes a new SATSolver object for a given number of variables.
func NewSATSolver(numVariables int) *SATSolver {
	graph := NewGraph(2 * numVariables) // Create a graph with 2 * numVariables nodes (positive and negative literals).
	return &SATSolver{
		graph:        graph,
		visited:      make(map[int]bool),
		visitedInv:   make(map[int]bool),
		scc:          make(map[int]int),
		numVariables: numVariables,
	}
}

// AddClause adds a clause to the implication graph.
// Each clause is represented as (a OR b), which is converted into two implications: NOT(a) → b and NOT(b) → a.
func (solver *SATSolver) AddClause(a, b int) {
	n := solver.numVariables
	if a > 0 && b > 0 {
		solver.graph.AddEdge(a+n, b) // NOT(a) → b.
		solver.graph.AddEdge(b+n, a) // NOT(b) → a.
	} else if a > 0 && b < 0 {
		solver.graph.AddEdge(a+n, -b+n) // NOT(a) → NOT(b).
		solver.graph.AddEdge(-b, a)     // b → a.
	} else if a < 0 && b > 0 {
		solver.graph.AddEdge(-a, b)     // a → b.
		solver.graph.AddEdge(b+n, -a+n) // NOT(b) → NOT(a).
	} else {
		solver.graph.AddEdge(-a, -b+n) // a → NOT(b).
		solver.graph.AddEdge(-b, -a+n) // b → NOT(a).
	}
}

// dfs performs a depth-first search on the original graph.
// It is used in the first pass of Kosaraju's algorithm to determine the order of nodes.
func (solver *SATSolver) dfs(node int) {
	if solver.visited[node] {
		return
	}
	solver.visited[node] = true
	for _, neighbor := range solver.graph.adj[node] {
		solver.dfs(neighbor)
	}
	solver.stack = append(solver.stack, node) // Push the node onto the stack after processing all its neighbors.
}

// dfsInv performs a depth-first search on the reversed graph.
// It is used in the second pass of Kosaraju's algorithm to identify SCCs.
func (solver *SATSolver) dfsInv(node int) {
	if solver.visitedInv[node] {
		return
	}
	solver.visitedInv[node] = true
	solver.scc[node] = solver.counter // Assign the current SCC ID to the node.
	for _, neighbor := range solver.graph.adjInv[node] {
		solver.dfsInv(neighbor)
	}
}

// Solve determines if the 2-SAT problem is satisfiable.
// It uses Kosaraju's algorithm to find SCCs and checks for contradictions.
func (solver *SATSolver) Solve() bool {
	// First pass: Perform DFS on the original graph to determine the finishing order of nodes.
	for i := 1; i <= 2*solver.numVariables; i++ {
		if !solver.visited[i] {
			solver.dfs(i)
		}
	}

	// Second pass: Perform DFS on the reversed graph to identify SCCs.
	for len(solver.stack) > 0 {
		node := solver.stack[len(solver.stack)-1]
		solver.stack = solver.stack[:len(solver.stack)-1]
		if !solver.visitedInv[node] {
			solver.counter++ // Increment the SCC ID counter.
			solver.dfsInv(node)
		}
	}

	// Check for contradictions: a variable and its negation must not be in the same SCC.
	for i := 1; i <= solver.numVariables; i++ {
		if solver.scc[i] == solver.scc[i+solver.numVariables] {
			return false // Unsatisfiable.
		}
	}
	return true // Satisfiable.
}

// ParseFile reads the input file and constructs a SATSolver object.
// The file contains the number of variables on the first line, followed by clauses on subsequent lines.
func ParseFile(filePath string) (*SATSolver, error) {
	file, err := os.Open(filePath) // Open the file.
	if err != nil {
		return nil, err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	// Read the number of variables.
	var numVariables int
	if scanner.Scan() {
		numVariables, err = strconv.Atoi(scanner.Text())
		if err != nil {
			return nil, fmt.Errorf("invalid variable count: %w", err)
		}
	}

	solver := NewSATSolver(numVariables)

	// Read each clause and add it to the solver.
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Fields(line)
		if len(parts) != 2 {
			continue // Skip invalid lines.
		}

		a, _ := strconv.Atoi(parts[0])
		b, _ := strconv.Atoi(parts[1])
		solver.AddClause(a, b)
	}

	return solver, nil
}

func main() {
	// Base path and configuration for input files.
	basePath := "course_4/module_4/programming_assignment_4/"
	filePrefix := "2sat"
	fileExtension := ".txt"
	fileCount := 6

	// Generate file paths for all instances.
	instances := make([]string, fileCount)
	for i := 1; i <= fileCount; i++ {
		instances[i-1] = fmt.Sprintf("%s%s%d%s", basePath, filePrefix, i, fileExtension)
	}

	var result string

	// Process each file and determine satisfiability.
	for _, filePath := range instances {
		log.Printf("Processing file: %s", filePath)

		solver, err := ParseFile(filePath)
		if err != nil {
			log.Printf("Error reading file %s: %v", filePath, err)
			result += "0" // Mark as unsatisfiable in case of an error.
			continue
		}

		if solver.Solve() {
			result += "1" // Instance is satisfiable.
		} else {
			result += "0" // Instance is unsatisfiable.
		}
	}

	// Print the final result as a binary string.
	fmt.Println("Result:", result)
}

File: course_4/module_4/programming_assignment_4/solution.py
class Graph:
    """
    Represents a directed graph for the 2-SAT problem.
    Contains adjacency lists for the original and reversed graphs.
    """
    def __init__(self, num_nodes):
        self.num_nodes = num_nodes
        self.adj = {i: [] for i in range(1, num_nodes + 1)}  # Original graph.
        self.adj_inv = {i: [] for i in range(1, num_nodes + 1)}  # Reversed graph.

    def add_edge(self, u, v):
        """
        Adds a directed edge from u to v in the original graph and its reverse in the reversed graph.
        """
        self.adj[u].append(v)
        self.adj_inv[v].append(u)


class SATSolver:
    """
    Encapsulates the logic for solving the 2-SAT problem using Kosaraju's algorithm.
    """
    def __init__(self, num_variables):
        self.num_variables = num_variables
        self.graph = Graph(2 * num_variables)  # 2 * num_variables for positive and negative literals.
        self.visited = set()  # Visited nodes for the original graph.
        self.visited_inv = set()  # Visited nodes for the reversed graph.
        self.stack = []  # Stack for Kosaraju's first pass.
        self.scc = {}  # Strongly Connected Components (SCC) mapping.
        self.counter = 0  # SCC counter.

    def add_clause(self, a, b):
        """
        Adds a clause to the implication graph.
        Each clause (a OR b) is converted to two implications: NOT(a) → b and NOT(b) → a.
        """
        n = self.num_variables
        if a > 0 and b > 0:
            self.graph.add_edge(a + n, b)  # NOT(a) → b
            self.graph.add_edge(b + n, a)  # NOT(b) → a
        elif a > 0 and b < 0:
            self.graph.add_edge(a + n, -b + n)  # NOT(a) → NOT(b)
            self.graph.add_edge(-b, a)  # b → a
        elif a < 0 and b > 0:
            self.graph.add_edge(-a, b)  # a → b
            self.graph.add_edge(b + n, -a + n)  # NOT(b) → NOT(a)
        else:
            self.graph.add_edge(-a, -b + n)  # a → NOT(b)
            self.graph.add_edge(-b, -a + n)  # b → NOT(a)

    def _dfs(self, node):
        """
        Depth-first search on the original graph for Kosaraju's first pass.
        """
        if node in self.visited:
            return
        self.visited.add(node)
        for neighbor in self.graph.adj[node]:
            self._dfs(neighbor)
        self.stack.append(node)  # Push the node to the stack after visiting all its neighbors.

    def _dfs_inv(self, node):
        """
        Depth-first search on the reversed graph for Kosaraju's second pass.
        """
        if node in self.visited_inv:
            return
        self.visited_inv.add(node)
        self.scc[node] = self.counter
        for neighbor in self.graph.adj_inv[node]:
            self._dfs_inv(neighbor)

    def solve(self):
        """
        Determines if the 2-SAT problem is satisfiable using Kosaraju's algorithm.
        """
        # First pass: Perform DFS on the original graph to determine the finishing order.
        for i in range(1, 2 * self.num_variables + 1):
            if i not in self.visited:
                self._dfs(i)

        # Second pass: Perform DFS on the reversed graph to find SCCs.
        while self.stack:
            node = self.stack.pop()
            if node not in self.visited_inv:
                self.counter += 1  # Increment SCC counter.
                self._dfs_inv(node)

        # Check for contradictions: a variable and its negation must not be in the same SCC.
        for i in range(1, self.num_variables + 1):
            if self.scc.get(i) == self.scc.get(i + self.num_variables):
                return False  # Unsatisfiable.
        return True  # Satisfiable.


def parse_file(file_path):
    """
    Reads the input file and constructs a SATSolver object with the clauses from the file.
    """
    with open(file_path, 'r') as file:
        num_variables = int(file.readline().strip())  # Read the number of variables.
        solver = SATSolver(num_variables)

        for line in file:
            a, b = map(int, line.split())
            solver.add_clause(a, b)

    return solver


def main():
    """
    Main function to process all problem instances and determine their satisfiability.
    """
    base_path = ""
    file_prefix = "2sat"
    file_extension = ".txt"
    file_count = 6

    # Generate file paths.
    instances = [f"{base_path}{file_prefix}{i}{file_extension}" for i in range(1, file_count + 1)]

    result = ""

    # Process each instance and determine satisfiability.
    for file_path in instances:
        print(f"Processing file: {file_path}")
        solver = parse_file(file_path)

        if solver.solve():
            result += "1"  # Instance is satisfiable.
        else:
            result += "0"  # Instance is unsatisfiable.

    # Print the final result as a binary string.
    print("Result:", result)


if __name__ == "__main__":
    main()
